### ISSUE TITLE
dropTarget {} (all) holders don't shuffle cards

### ISSUE THREAD

**Issue by bjalder26** [ArnoldSmith86/virtualtabletop#596]
If a holder's dropTarget is set to {} (all) then it will not shuffle the cards, instead it appears to sort them.  I am not really sure why.  However, I do know it also shuffles the shuffle button and the pile.  The shuffle button and pile appear to actually get shuffled.  However, the cards just seem to go up higher and higher in Z value while also becoming sorted.  Even if I pre-shuffle the cards before setting the dropTarget to {}, they appear to get sorted.

This room will demonstrate this issue.
https://virtualtabletop.io/6e07

**Comment by 96LawDawg** [comment]
This is a very significant bug.  Hopefully someone will look at it soon.

**Comment by RaphaelAlvez** [comment]
At First look this seems like a problem related to piles. Maybe this can be tested by adding piles to the dropTarget

**Comment by bjalder26** [comment]
> At First look this seems like a problem related to piles. Maybe this can be tested by adding piles to the dropTarget

Good idea, here is what I found:  I took 3 cards and the pile they are on and labeled them all with the same property, then made that property the dropTarget, and they shuffled fine.  Then I removed it from the pile and they still shuffled fine.  Then I applied it to the button, and they stopped shuffling and started sorting.  So it appears to be an issue with the button getting shuffled.

**Comment by ArnoldSmith86** [comment]
The problem is the function `holder.children()` which returns the children of the holder the game author cares about - in theory.

The logic is this:

1. Get all children of the holder that match its `dropTarget` definition or are `type: pile`.
2. If it is exactly one child and it is `type: pile`, instead get all children of that pile that match the `dropTarget`.

So if you set `dropTarget: {}` and there's a button and a pile in the holder, those two will be the children that get shuffled or sorted or whatever.

The question is: how is the engine supposed to know that you didn't want to shuffle the button?

**Comment by bjalder26** [comment]
> and there's a button and a pile in the holder, those two will be the children that get shuffled or sorted or whatever.

Maybe it should just put all the pile children and other widgets in a collection and shuffle them.  I don't think it actually matters that it shuffles the button (and deck) too.  If I understand the issue correctly, maybe the real issue is in that specific situation the pile's children aren't being passed into the collection instead of the pile.

**Comment by robartsd** [comment]
What should happen in a holder with `alignChildren:false` that contains multiple piles? Should the cards in each pile be shuffled (together or reach pile independently)? As a game author, I don't usually care about the piles (and shuffle on holders without alignChildren doesn't make sense to me), I think I'd want each pile in the holder shuffled independently.

**Comment by bjalder26** [comment]
> I think I'd want each pile in the holder shuffled independently.

Hmm... this sounds like a good question.  If you want cards shuffled together, then you could make them 1 deck, and if you wanted cards shuffled separately then you could put them in two different holders.  So I guess we're talking about:
1) What if someone wants two decks to be separate sometimes, and shuffled together at other times?
2) What if someone wants two separate decks to be held in the same holder, but shuffled separately.

Currently the workaround for 1) would be to not have the decks in the same holder, but to select cards by deck for the 2nd deck, then change parent to move the second deck's cards into the holder before shuffling. 
Currently, the workaround for 2) I suppose would be to shuffle the decks separately, then move the 2nd deck into the holder after the shuffle.

I'm not really sure which situation is more likely or more of an issue.

**Comment by robartsd** [comment]
You can always put cards for different decks into the same pile then shuffle them.

**Comment by bjalder26** [comment]
> You can always put cards for different decks into the same pile then shuffle them.

Yeah, I was thinking about this wrong.  I agree that in that situation I would expect them to not all get shuffled together.

**Comment by bjalder26** [comment]
I did some tracing for this issue in case it is helpful. 

`Clicked shuffle and didn't shuffle cards in dropTarget {} holder`
